include/swift/AST/Module.h

```diff
@@ -165,27 +165,36 @@ class ModuleSourceFileLocationMap;
 class PackageUnit: public DeclContext {
 
   Identifier PackageName;
+  ModuleDecl &SourceModule;
 
-  PackageUnit(Identifier name);
+  PackageUnit(Identifier name, ModuleDecl &src)
+    : DeclContext(DeclContextKind::Package, nullptr), PackageName(name), SourceModule(src) {}
 
 public:
   static PackageUnit *
-  create(Identifier name, ASTContext &ctx) {
-    return new (ctx) PackageUnit(name);
+  create(Identifier name, ModuleDecl &src, ASTContext &ctx) {
+    return new (ctx) PackageUnit(name, src);
```

Comment:
This is…pretty weird. Wouldn’t we prefer to have one instance of `PackageUnit` shared by all of the `ModuleDecl`s in the same package? That way you could `==` the DeclContext pointers to see if they match, just like other DeclContexts.

Reply:
I found this way easier to look up the module from a package unit in `DeclContext::getModuleScopeContext` and `DeclContext::getParentModule`; otherwise will have to compare ModuleDecl pointers in an array

Reply:
That kind of just moves the question. Why do `getParentModule()` and `getModuleScopeContext()` return the module when you call them on the package? That seems backwards from what you’d want.

Reply:
The existing code assumes ModuleDecl is the top level context as long as there's non-null DeclContext (otherwise `public`) and performs `isa`/`cast` without checking if it's null, which is everywhere. With a package context now introduced, those existing functions are still called on the target DeclContext that may or may not be package, and will crash on `isa`/`cast` if the returned context is not module. Either we'll have to add assert or null check before casting or add package specific handling before each; by returning the module mapped to a package as in this PR, the changes required are minimal.

Reply:
Also updated doc comments in AccessScope and DeclContext

Reply:
I tried adding a PackageUnit-ModuleDecl map in `ASTContext`, but that didn't seem to work, because `DeclContext::getASTContext()` calls `getParentModule()`, and using the map in `ASTContext` to look up the module inside `getParentModule()` and results in a cycle, so resorted to having a weak reference to `ModuleDecl` from `PackageUnit`. Lmk if if the way it references is not actually a weak reference though.